### DISCUSSION TITLE
Language localization compatibility

### DISCUSSION THREAD

**Discussion by atakanau** [codeigniter4/shield#674] [Q&A] [answered]
I am working on language localization. I am using "Shield" as "Third Party".

I prepare the global variables in:
`\app\Controllers\BaseController.php`
``` php
public function initController(RequestInterface $request, ...)
{
	// Do Not Edit This Line
	parent::initController($request, $response, $logger);
	// Preload any models, libraries, etc, here.
	// E.g.: $this->session = \Config\Services::session();
	$this->helpers = array_merge($this->helpers, ['setting']);
	
	// localization >>>
	$this->viewData['locale'] = [];
	$this->viewData['locale']['current_code'] = $request->getLocale();
	$this->viewData['locale']['codes'] = $request->config->supportedLocales;
	$this->viewData['locale']['lang_switcher'] = array(
		 'en'=>[ 'html_lang'=>'en-US'
			,'html_dir'=>'ltr'
			,'codename'=>'English'
		]
		,'es'=>[ 'html_lang'=>'es-ES'
			,'html_dir'=>'ltr'
			,'codename'=>'Español'
		]
	);
	foreach($request->config->supportedLocales as $code){
		// Other language variable values ​​etc.
```
I discovered following setting the for URL routing:
`\app\ThirdParty\shield\src\Config\AuthRoutes.php`
``` php
    public array $routes = [
        'register' => [
            [
                'get',
                '{locale}/register', // <- old value: 'register'
```

In this situation; `[locale]/login`, `[locale]/login/magic-link` and `[locale]/register` works.

I have activated the E-Mail Activation. I had to make the following new setting for redirect by page (url) language:
`\app\ThirdParty\shield\src\Controllers\RegisterController.php`
``` php
// If an action has been defined for register, start it up.
$hasAction = $authenticator->startUpAction('register', $user);
if ($hasAction) {
    return redirect()->to($this->request->getLocale().'/auth/a/show');
    // return redirect()->to('auth/a/show'); // <- old
}
```
However, I can't get to my `viewData` variable before the page at `[locale]/auth/a/show` is displayed.
`\app\ThirdParty\shield\src\Traits\Viewable.php`
```php
protected function view(string $view, array $data = [], array $options = []): string
{
    if( isset($this->viewData) ) // [locale]/register ok ; fails on [locale]/auth/a/show
        $data = array_merge($this->viewData, $data);
    return view($view, $data, $options);
}
```
I need to edit the action address of the form. I need access to current language code (and other variables).
`\app\ThirdParty\shield\src\Views\email_activate_show.php`
```html
<form action="<?= site_url('auth/a/verify') ?>" method="post">
```
```html
<form action="<?= site_url($locale['current_code'].'/auth/a/verify') ?>" method="post">
```
How can I access the `viewData` variable in all "views"?

**Comment by kenjis** [comment]
> How can I access the viewData variable in all "views"?

Pass the variable to views.
```php
view($view, ['viewData' => $this->viewData], $options);
```

**Reply by atakanau** [reply]
> > How can I access the viewData variable in all "views"?
> 
> Pass the variable to views.
> 
> ```
> view($view, ['viewData' => $this->viewData], $options);
> ```
Hi @kenjis , thanks for your reply. I can use  `$this->viewData` when the url is `[locale]/login`, `[locale]/login/magic-link` and `[locale]/register`. But `$this->viewData` is **undefined** when the url is `[locale]/auth/a/show` (eg. `http://localhost/ci_4/public_html/en/auth/a/show`):
### ErrorException
#### Undefined property: CodeIgniter\Shield\Authentication\Actions\EmailActivator::$viewData
##### APPPATH\ThirdParty\shield-dev\src\Traits\Viewable.php at line 16
```php
14     protected function view(string $view, array $data = [], array $options = []): string
15     {
16         return view($view, ['viewData' => $this->viewData], $options);
17     }
18 
```

**Reply by kenjis** [reply]
`CodeIgniter\Shield\Authentication\Actions\EmailActivator` is not a Controller.
So there is no `$this->viewData`.

**Reply by atakanau** [reply]
😔 Ok. 
As I mentioned earlier,briefly:
I prepare my variable in `\app\Controllers\BaseController.php`:
`protected $viewData = [];`
I can access the `$this->viewData` in `APPPATH\ThirdParty\shield-dev\src\Traits\Viewable.php` when the url is `[locale]/login`, `[locale]/login/magic-link` and `[locale]/register`. But `$this->viewData` is **undefined** when the url is `[locale]/auth/a/show`.
I need to edit the action address of the form. I need access to current language code (and my other variables).

**How can I access the viewData variable in all "views"?**

**Reply by kenjis** [reply]
You cannot access `$viewData` in a view unless you pass the variable to the view.
`CodeIgniter\Shield\Authentication\Actions\EmailActivator` is not a Controller.
Read the `ActionController` code.

**Reply by atakanau** [reply]
I did not edit the `CodeIgniter\Shield\Authentication\Actions\EmailActivator` file. Please read that error again. As you say:
> Pass the variable to views.
> 
> ```
> view($view, ['viewData' => $this->viewData], $options);
> ```
I modified the `APPPATH\ThirdParty\shield-dev\src\Traits\Viewable.php`; I added the `$this->viewData` as a parameter to the view method. Controller calls this view method. But `$this->viewData` is **undefined**. So, result: ErrorException. ( Again: I can access the `$this->viewData` in `APPPATH\ThirdParty\shield-dev\src\Traits\Viewable.php` when the url is `[locale]/login`, `[locale]/login/magic-link` or `[locale]/register`. Same code fails on `[locale]/auth/a/show` ! )

Anyway,
> Read the ActionController code.

I am accessing the `$this->viewData` inside the show and verify methods in `shield-dev\src\Controllers\ActionController.php`. How do I pass `$this->viewData` to view? Sorry, I'm confused.

**Reply by kenjis** [reply]
This is just an example. Don't see it as a correct answer for your case.
```php
view($view, ['viewData' => $this->viewData], $options);
```

**Reply by kenjis** [reply]
It seems you don't understand what is a class in PHP.

`$this->viewData` is a property in the instance.

> Undefined property: CodeIgniter\Shield\Authentication\Actions\EmailActivator::$viewData

You do not define the property `$viewData` in `EmailActivator`.
So there is no wonder the error occurs.

**Comment by kenjis** [comment, marked as answer]
I recommend you set the viewData to the View instance with `setVar()` in your controller.
See https://codeigniter4.github.io/CodeIgniter4/outgoing/view_renderer.html

**Reply by atakanau** [reply]
Thank you. `viewData()`  is worked. 👌
`\app\Controllers\BaseController.php`:
```
  // localization >>>
  $this->viewData['locale'] = [];
  $this->viewData['locale']['current_code'] = $request->getLocale();
  $this->viewData['locale']['codes'] = $request->config->supportedLocales;
  // other custom values...
  $view = \Config\Services::renderer();
  $view->setData($this->viewData);
```
